### DISCUSSION TITLE
Cache + redirect have inconsistent behavior on deployed environment

### DISCUSSION THREAD

**Discussion by ababol** [vercel/next.js#24537] [Help]
### What version of Next.js are you using?

10.1.3

### What version of Node.js are you using?

14

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

👋  I build a customise localeDetection on my website to handle subpage (not only the root one).

Here is the [small MVP project](https://github.com/shotgun-warehouse/cache-redirect-issue/blob/master/pages/this-page-was-cached-in-en.js#L35).

So far so good, it's an easy script which redirect the user according to the browser/headers locale.

But when adding `Cache-Control` and caching the page first with the `defaultLocale` (`en` in this case), it doesn't work, the redirect script is just ignored.

LocaleDetection + Cache are working fine on root page `/` so I am wondering if there is an exception or something on vercel to allow custom getInitialProps/getServerSideProps on root page but not on subpage? I would be really happy to resolve this issue, since this is a blocking one to migrate my website :(.

### Expected Behavior

Redirect script should work even with cache-control on and if the page was first cached in english (defaultLocale).

### To Reproduce

- Go to [this link](https://cache-redirect-issue-ababol-shotgun-warehouse.vercel.app/this-page-was-cached-in-en) with a FR browser. You will not be redirected since the page was first cache with an en browser/default locale one.

- But if you go to [this link](https://cache-redirect-issue-ababol-shotgun-warehouse.vercel.app/this-page-was-cached-in-fr), it's working fine since the page was first cached first with a FR browser.

Meaning behavior is inconsistent depending on how the page was first cached. :( The second page behavior is the expected one.
The behavior should be the same even if `Cache-Control` is [on every page like in this small project](https://github.com/shotgun-warehouse/cache-redirect-issue/blob/master/next.config.js#L12-L18).

**Comment by ijjk** [comment]
Hi, it looks like the `cache-control` header you are setting [here](https://github.com/shotgun-warehouse/cache-redirect-issue/blob/89be28672436cd6b63acb7f6576180850a32d224/next.config.js#L11-L19) is overriding the redirect behavior. If you would like to use revalidate with i18n routing you should using `getStaticProps` with the `revalidate` feature instead. 

Note: currently the `cache-control` header is being applied to all paths and assets which can have un-intended side-affects as it should only be applied to specific assets you meant to have that caching behavior applied to. For example, this caching behavior for pages should typically be done through `getStaticProps` `revalidate` or setting the header in `getServerSideProps` instead to keep it at the page level.
